src/main/java/io/temporal/samples/encryptedpayloads/CryptDataConverter.java

```diff
@@ -0,0 +1,189 @@
+ */
+
+package io.temporal.samples.encryptedpayloads;
+
+import com.google.common.base.Defaults;
+import com.google.protobuf.ByteString;
+import io.temporal.api.common.v1.Payload;
+import io.temporal.api.common.v1.Payloads;
+import io.temporal.common.converter.DataConverter;
+import io.temporal.common.converter.DataConverterException;
+import java.lang.reflect.Type;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.security.SecureRandom;
+import java.util.Optional;
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+class EncodingKeys {
+  static final String METADATA_ENCODING_KEY = "encoding";
+  static final String METADATA_WRAPPED_ENCODING_KEY = "wrapped-encoding";
+
+  static final String METADATA_ENCODING_RAW_NAME = "binary/plain";
+  static final ByteString METADATA_ENCODING_RAW = ByteString.copyFrom(METADATA_ENCODING_RAW_NAME,
+      StandardCharsets.UTF_8);
+}
+
+public class CryptDataConverter implements DataConverter {
+  private static final String CIPHER = "AES/GCM/NoPadding";
+  private static final int GCM_NONCE_LENGTH_BYTE = 12;
+  private static final int GCM_TAG_LENGTH_BIT = 128;
+  private static final Charset UTF_8 = StandardCharsets.UTF_8;
+
+  private DataConverter converter;
+
+  public CryptDataConverter(DataConverter converter) {
+    this.converter = converter;
+  }
+
+  private SecretKey getKey() {
+    // Key can be fetched from KMS or other secure storage.
+    return new SecretKeySpec("test-key-test-key-test-key-test!".getBytes(UTF_8), "AES");
+  }
+
+  private static byte[] getNonce(int size) {
+    byte[] nonce = new byte[size];
+    new SecureRandom().nextBytes(nonce);
+    return nonce;
+  }
+
+  private byte[] encrypt(byte[] plainData) throws Exception {
+    SecretKey key = this.getKey();
+    byte[] nonce = getNonce(GCM_NONCE_LENGTH_BYTE);
+
+    Cipher cipher = Cipher.getInstance(CIPHER);
+    cipher.init(Cipher.ENCRYPT_MODE, key, new GCMParameterSpec(GCM_TAG_LENGTH_BIT, nonce));
+
+    byte[] encryptedData = cipher.doFinal(plainData);
+    byte[] result = ByteBuffer.allocate(nonce.length + encryptedData.length).put(nonce).put(encryptedData).array();
+
+    return result;
+  }
+
+  private byte[] decrypt(byte[] encryptedDataWithNonce) throws Exception {
+    SecretKey key = this.getKey();
+    ByteBuffer buffer = ByteBuffer.wrap(encryptedDataWithNonce);
+
+    byte[] nonce = new byte[GCM_NONCE_LENGTH_BYTE];
+    buffer.get(nonce);
+    byte[] encryptedData = new byte[buffer.remaining()];
+    buffer.get(encryptedData);
+
+    Cipher cipher = Cipher.getInstance(CIPHER);
+    cipher.init(Cipher.DECRYPT_MODE, key, new GCMParameterSpec(GCM_TAG_LENGTH_BIT, nonce));
+
+    return cipher.doFinal(encryptedData);
+  }
+
+  @Override
+  public <T> Optional<Payload> toPayload(T value) throws DataConverterException {
```

Comment:
I don't think we want to encrypt data passed to toPayload. This method is used for searchable attributes and header fields. I don't think they should be encrypted.

Reply:
Makes sense, fixed.